src/Workspaces/Core/Portable/Workspace/Solution/TextLoader.cs

```diff
@@ -13,22 +18,120 @@ namespace Microsoft.CodeAnalysis
         /// </summary>
+        /// <exception cref="IOException" />
+        /// <exception cref="InvalidDataException"/>
+        /// <exception cref="OperationCanceledException"/>
         internal virtual TextAndVersion LoadTextAndVersionSynchronously(Workspace workspace, DocumentId documentId, CancellationToken cancellationToken)
         {
             // this implementation exists in case a custom derived type does not have access to internals
             return LoadTextAndVersionAsync(workspace, documentId, cancellationToken).WaitAndGetResult_CanCallOnBackground(cancellationToken);
         }
 
+        internal async Task<TextAndVersion> LoadTextAsync(Workspace workspace, DocumentId documentId, CancellationToken cancellationToken)
+        {
+            var retries = 0;
+
+            while (true)
+            {
+                try
+                {
+                    return await LoadTextAndVersionAsync(workspace, documentId, cancellationToken).ConfigureAwait(continueOnCapturedContext: false);
+                }
+                catch (IOException e)
+                {
+                    if (++retries > MaxRetries)
+                    {
+                        return CreateFailedText(workspace, documentId, e.Message);
+                    }
+
+                    // fall out to try again
+                }
+                catch (InvalidDataException e)
+                {
+                    return CreateFailedText(workspace, documentId, e.Message);
+                }
+
+                // try again after a delay
+                await Task.Delay(RetryDelay, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        internal TextAndVersion LoadTextSynchronously(Workspace workspace, DocumentId documentId, CancellationToken cancellationToken)
+        {
+            var retries = 0;
+
+            while (true)
+            {
+                try
+                {
+                    return LoadTextAndVersionSynchronously(workspace, documentId, cancellationToken);
+                }
+                catch (IOException e)
+                {
+                    if (++retries > MaxRetries)
+                    {
+                        return CreateFailedText(workspace, documentId, e.Message);
+                    }
+
+                    // fall out to try again
+                }
+                catch (InvalidDataException e)
+                {
+                    return CreateFailedText(workspace, documentId, e.Message);
+                }
+
+                // try again after a delay
+                Thread.Sleep(RetryDelay);
+            }
+        }
+
+        private TextAndVersion CreateFailedText(Workspace workspace, DocumentId documentId, string message)
+        {
+            // Notify workspace for backwards compatibility.
+            workspace.OnWorkspaceFailed(new DocumentDiagnostic(WorkspaceDiagnosticKind.Failure, message, documentId));
```

Comment:
❓ Is there a way we can move this later? In other words, have the workspace check for a failure diagnostic(s) and call `OnWorkspaceFailed` if any exist?

Reply:
When would we check?